### PR TITLE
[Integrate] As a user, I can see the list of surveys

### LINKED ISSUE #17: [Integrate] As a user, I can see the list of surveys
## Why

App must call backend API to fetch list of surveys

## Acceptance Criteria

- Make post request to `https://survey-api.nimblehq.co/api/v1/surveys` and handle response
- If successful, show the list of surveys
- if Unsuccessful, show a toast with the error from API
- on pull to refresh, show a loading indicator while fetching surveys
- fetch next page of surveys before swiping to the last item

## Resources

- [API](https://documenter.getpostman.com/view/11835486/UVsPNPfU#064357a5-4625-4ef7-b042-0d1f7e2f2acb)

### PR DESCRIPTION
**Pull request by ryan-conway**
- closes #17 

## What happened 👀

- Created `HomeViewModel` and `HomeState`
- Integrated `HomeViewModel` with `HomeScreen`
- Added rxdart dependency
- Fixed issue with java version

## Insight 📝

- Project kept failing to build due to mismatched java version, to fix this we need to apply `org.gradle.jvmargs=--add-opens java.base/java.io=ALL-UNNAMED` to `gradle.properties` 🥴 

## Proof Of Work 📹

https://github.com/nimblehq/flutter-ic-lydia-ryan/assets/53168251/b2dd8e9b-eae1-4dcc-bf2b-cbbc4060e8fb

### CODE REVIEW

File: lib/screens/home/home_view_model.dart [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  final BehaviorSubject<List<SurveyUiModel>> _surveys = BehaviorSubject();
+  Stream<List<SurveyUiModel>> get surveys => _surveys.stream;
+
+  final BehaviorSubject<String> _error = BehaviorSubject();
+  Stream<String> get error => _error.stream;
+
+  HomeViewModel(this._getSurveysUseCase) : super(const HomeState.init());
+
+  void getSurveys() async {
+    final result = await _getSurveysUseCase.call();
```
Comment (markgravity) [inline lib/screens/home/home_view_model.dart:18]: Do we miss updating the state to `HomeState.loading()` here? 🤔

File: lib/screens/home/home_screen.dart [hunk 2]
```diff
@@ -11,53 +29,32 @@ class HomeScreen extends ConsumerStatefulWidget {
...
-  )
-];
+Widget _buildHomeScreen(
+  List<SurveyUiModel> surveys,
+  bool loading,
+  String? errorMessage,
+) {
+  return Scaffold(
+    body: HomePagingWidget(surveys: surveys),
+  );
```
Comment (manh-t) [inline lib/screens/home/home_screen.dart]: @ryan-conway should we handle loading/ error too? 😅

File: lib/screens/home/home_screen.dart [hunk 3]
```diff
@@ -11,53 +29,32 @@ class HomeScreen extends ConsumerStatefulWidget {
...
-    return Scaffold(
-      body: HomePagingWidget(surveys: _mockSurveyUiModels),
-    );
+    final surveys = ref.watch(_surveysStreamProvider).value ?? [];
+    final error = ref.watch(_errorStreamProvider).value;
+
+    return ref.watch<HomeState>(_homeViewModelProvider).when(
+          init: () => const HomeLoadingWidget(),
+          success: () => _buildHomeScreen(surveys, false, null),
+          loading: () => _buildHomeScreen(surveys, true, null),
```
Comment (manh-t) [inline lib/screens/home/home_screen.dart]: @ryan-conway just curious, what is the different between init and loading states? 💭

Comment (ryan-conway) [reply lib/screens/home/home_screen.dart]: `init` should be for the initial loading state with shimmer, and `loading` was planned to be for loading the next page of surveys but this is not yet implemented 🙈 Shall we implement pagination in another task instead, what do you think? CC: @markgravity

File: lib/screens/home/home_view_model.dart (hunk 1 above)
Comment (ryan-conway) [reply lib/screens/home/home_view_model.dart:18]: In the first implementation, setting the state to `loading` immediately caused the app to crash 🙈 Currently I have removed the loading state and just use the `init` state to show the loading shimmer 🙏🏻

File: lib/screens/home/home_screen.dart (hunk 3 above)
Comment (manh-t) [reply lib/screens/home/home_screen.dart]: @ryan-conway yeah, I think we can implement pagination in another task 👍

File: lib/screens/home/home_view_model.dart [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import 'package:rxdart/subjects.dart';
+
+class HomeViewModel extends StateNotifier<HomeState> {
+  final GetSurveysUseCase _getSurveysUseCase;
+
+  final BehaviorSubject<List<SurveyUiModel>> _surveys = BehaviorSubject();
+  Stream<List<SurveyUiModel>> get surveys => _surveys.stream;
+
+  final BehaviorSubject<String> _error = BehaviorSubject();
+  Stream<String> get error => _error.stream;
```
Comment (manh-t) [inline lib/screens/home/home_view_model.dart]: we can use `error` from `HomeState` and get rid of this behavior. Tips: We can listen to the update of `HomeState` by calling `ref.listen` ([reference](https://riverpod.dev/docs/concepts/reading#using-reflisten-to-react-to-a-provider-change))

**Review by manh-t — COMMENTED** [review]
rest LGTM

File: lib/screens/home/home_view_model.dart (hunk 4 above)
Comment (ryan-conway) [reply lib/screens/home/home_view_model.dart]: Updated in [a57ad5d](https://github.com/nimblehq/flutter-ic-lydia-ryan/pull/79/commits/a57ad5d858ad34ba1c35494ecbac1a8332a45af2) 🙏🏻

**Review by markgravity — APPROVED** [review]

File: lib/screens/home/home_view_model.dart [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+class HomeViewModel extends StateNotifier<HomeState> {
+  final GetSurveysUseCase _getSurveysUseCase;
+
+  final BehaviorSubject<List<SurveyUiModel>> _surveys = BehaviorSubject();
+  Stream<List<SurveyUiModel>> get surveys => _surveys.stream;
+
+  HomeViewModel(this._getSurveysUseCase) : super(const HomeState.init());
+
+  Future<void> getSurveys() async {
```
Comment (manh-t) [inline lib/screens/home/home_view_model.dart:17]: @ryan-conway should we write unittest for this VM?

Comment (ryan-conway) [reply lib/screens/home/home_view_model.dart:17]: @manh-t Unit test added in [1a98c09](https://github.com/nimblehq/flutter-ic-lydia-ryan/pull/79/commits/1a98c09797dc6bdaee5ae8d03bf49e6182cff9b4) 🙏🏻

**Review by manh-t — APPROVED** [review]

**Pull request merged**